### DISCUSSION TITLE
Allow API usage to capture logs in a string

### DISCUSSION THREAD

**Discussion by chahu** [sqlalchemy/alembic#1418] [Ideas]
I had the need to trigger alembic commands in one of my flask endpoints. One of the goals was to include the output from alembic in the returned json. For commands such as `upgrade()`, the entirety of what is output is done via logging. Seemed straight forward to add a handler to a standard python logger and store the logs. What I attempted was roughly:

```
import logging
from io import StringIO
from alembic import command
from flask import current_app

log_stream = StringIO()
handler = logging.StreamHandler(log_stream)
handler.setLevel(logging.INFO)
logger = logging.getLogger('alembic.runtime.migration')
logger.addHandler(handler)
migrate_config = current_app.extensions["migrate"].migrate.get_config()
command.upgrade(migrate_config, 'head')
logged_output = log_stream.getvalue() # nothing here, as my handler was removed when alembic used fileConfig()
```

The issue I discovered is configuring the python logger used by alembic proved impossible as "existing loggers are disabled" when the config file is read by `logging.config.fileConfig`. There is an optional argument `disable_existing_loggers` which would allow the above code snippet to work, but alembic doesn't pass it or provide any way to do so. I've given up on getting alembic's output and instead am generating my own, but wanted to report this limitation.

https://docs.python.org/3/library/logging.config.html#logging.config.fileConfig

**Comment by zzzeek** [comment]
that env.py is yours to modify freely.   take out the fileconfig and do whatever you want to set up logging.    alembic does nothing with loggers or handlers whatsoever except write messages to loggers, it has nothing to do with existing loggers or anything else like that, it's 100% that one line of code that is part of your own application to change in any way you'd like.

**Reply by chahu** [reply]
Thanks for the quick reply. I didn't realize this code was copied to the migrations directory and editable by the project. I appreciate the help.
